src/combinatorics/hall/basic.lean

```diff
@@ -0,0 +1,230 @@
+/-
+Copyright (c) 2021 Alena Gusakov, Bhavik Mehta, Kyle Miller. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Alena Gusakov, Bhavik Mehta, Kyle Miller
+-/
+import combinatorics.hall.finite
+import topology.category.Top.limits
+
+/-!
+# Hall's Marriage Theorem
+
+Given a list of finite subsets $X_1,X_2,\dots,X_n$ of some given set
+$S$, Hall in [Hall1935] gave a necessary and sufficient condition for
+there to be a list of distinct elements $x_1,x_2,\dots,x_n$ with
+$x_i\in X_i$ for each $i$: it is when for each $k$, the union of every
+$k$ of these subsets has at least $k$ elements.
+
+This file proves this for an indexed family `t : ι → finset α` of
+finite sets as well as some variants of the statement.  The list of
+distinct representatives is given by an injective function `f : ι → α`
+such that `∀ i, f i ∈ t i`.  The classic version of Hall's theorem
+is when `ι` is a `fintype`, but this module proves it with unconstrained
+cardinality.  The way we do this is (1) prove the `fintype` version
+in `combinatorics.hall.finite` and then (2) use the fact that inverse
+limits of nonempty finite sets are nonempty (`nonempty_sections_of_fintype_inverse_system`)
+to lift this result to the unconstrained case.
+
+A description of this formalization is in [Gusakov2021].
+
+## Main statements
+* `finset.all_card_le_bUnion_card_iff_exists_injective` is in terms of `t : ι → finset α`.
+* `fintype.all_card_le_rel_image_card_iff_exists_injective` is in terms of a relation
+  `r : α → β → Prop` such that `rel.image r {a}` is a finite set for all `a : α`.
+* `fintype.all_card_le_filter_rel_iff_exists_injective` is in terms of a relation
+  `r : α → β → Prop` on finite types, with the Hall condition given in terms of
+  `finset.univ.filter`.
+
+## Todo
+
+* The statement of the theorem in terms of bipartite graphs is in preparation.
+
+## Tags
+
+Hall's Marriage Theorem, indexed families
+-/
+
+open finset
+
+namespace hall_marriage_theorem
+
+instance finset.directed_order (α : Type*) : directed_order (finset α) :=
+⟨λ s t, by { classical, exact ⟨s ∪ t, subset_union_left s t, subset_union_right s t⟩ }⟩
+
+/-- The set of matchings for `t` when restricted to a `finset` of `ι`. -/
+def hall_matchings_on {ι α : Type*} (t : ι → finset α) (ι' : finset ι) :=
+{f : ι' → α | function.injective f ∧ ∀ x, f x ∈ t x}
```

Comment:
Is there any reason for this subtype? I feel like the following definition is more natural and would allow you to talk about the intersection of sets of matchings.
```suggestion
def hall_matchings_on {ι α : Type*} (t : ι → finset α) (s : finset ι) :=
{f : ι → α | (s : set ι).inj_on f ∧ ∀ x ∈ s, f x ∈ t x}
```

Reply:
Yes, it needs to be this way for `hall_matchings_on.fintype` to hold, and finiteness is essential for the compactness argument.  Subtypes give the correct extensionality property (no junk data outside the finset).

Reply:
I'm not really inclined to rewrite this already working (and somewhat fiddly) module to check whether that improves anything... If you have time, feel free to check and let me know.